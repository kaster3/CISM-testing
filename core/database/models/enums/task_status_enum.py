from enum import Enum


class TaskStatusEnum(Enum):
    NEW = "Новая задача"
    IN_PROGRESS = "В процессе работы"
    COMPLETED = "Завершено успешно"
    ERROR = "Ошибка"
