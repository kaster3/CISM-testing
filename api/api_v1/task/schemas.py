from pydantic import BaseModel, ConfigDict

from core.database.models.enums.task_status_enum import TaskStatusEnum


class TaskBaseSchema(BaseModel):
    name: str


class TaskSchema(TaskBaseSchema):
    model_config = ConfigDict(from_attributes=True)
    id: int
    status: TaskStatusEnum


class TaskCreate(TaskBaseSchema):
    status: TaskStatusEnum = TaskStatusEnum.NEW


class TaskUpdatePartial(TaskBaseSchema):
    name: str | None = None
    status: TaskStatusEnum | None = None
