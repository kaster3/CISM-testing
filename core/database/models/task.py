from sqlalchemy import Enum as SqlAlchemyEnum
from sqlalchemy import String
from sqlalchemy.orm import Mapped, mapped_column

from core.database.models.base import Base
from core.database.models.enums.task_status_enum import TaskStatusEnum
from core.database.models.mixins.pk_id_mixin import IntIdPkMixin


class Task(IntIdPkMixin, Base):
    name: Mapped[str] = mapped_column(String(20))
    status: Mapped[TaskStatusEnum] = mapped_column(
        SqlAlchemyEnum(TaskStatusEnum),
        default=TaskStatusEnum.NEW,
    )

    def __repr__(self) -> str:
        return f"Task(id={self.id}, name={self.name!r}, status={self.status})"
