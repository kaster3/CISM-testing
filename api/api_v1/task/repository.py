from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from api.api_v1.task.schemas import TaskCreate, TaskUpdatePartial
from core.database.models import Task
from core.database.models.enums.task_status_enum import TaskStatusEnum


class TaskRepository:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def get_task(self, task_id: int) -> Task | None:
        task = await self.session.get(Task, task_id)
        return task

    async def get_all_tasks(self) -> list[Task]:
        stmt = select(Task)
        tasks = await self.session.scalars(stmt)
        return list(tasks)

    async def get_tasks_by_status(self, task_status: TaskStatusEnum) -> list[Task]:
        stmt = select(Task).where(Task.status == task_status)
        tasks = await self.session.scalars(stmt)
        return list(tasks)

    async def create_task(self, task_in: TaskCreate) -> Task:
        task = Task(**task_in.model_dump())
        self.session.add(task)
        await self.session.commit()
        await self.session.refresh(task)
        return task

    async def update_task(
        self,
        task: Task,
        task_in: TaskUpdatePartial,
    ) -> Task:
        for name, value in task_in.model_dump(exclude_unset=True).items():
            setattr(task, name, value)
        await self.session.commit()
        return task
