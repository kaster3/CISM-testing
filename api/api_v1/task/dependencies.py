from typing import Annotated

from fastapi import Depends, HTTPException, Path
from sqlalchemy.ext.asyncio import AsyncSession

from api.api_v1.task.repository import TaskRepository
from api.api_v1.task.service import TaskService
from core.database.db_helper import db_helper
from core.database.models import Task


async def get_task_repository(
    session: AsyncSession = Depends(db_helper.session_getter),
) -> TaskRepository:
    return TaskRepository(session=session)


async def get_task_service(
    repo: Annotated[TaskRepository, Depends(get_task_repository)],
) -> TaskService:
    return TaskService(repo=repo)


async def get_task_by_id(
    repo: Annotated[TaskRepository, Depends(get_task_repository)],
    task_id: Annotated[int, Path],
) -> Task:
    task = await repo.get_task(task_id=task_id)
    if task is not None:
        return task
    raise HTTPException(status_code=404, detail="Task not found")
