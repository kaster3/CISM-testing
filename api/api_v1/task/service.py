import logging

from fastapi import HTTPException

from api.api_v1.task.repository import TaskRepository
from api.api_v1.task.schemas import TaskCreate
from core.database.models import Task
from core.database.models.enums.task_status_enum import TaskStatusEnum

log = logging.getLogger(__name__)


class TaskService:
    def __init__(self, repo: TaskRepository) -> None:
        self.repo = repo


    async def get_tasks(self, task_status: TaskStatusEnum | None) -> list[Task]:
        if task_status is None:
            tasks = await self.repo.get_all_tasks()
        else:
            tasks = await self.repo.get_tasks_by_status(task_status=task_status)
        if not tasks:
            raise HTTPException(status_code=404, detail="Tasks not found")
        return tasks

    async def create_task(self, task_in: TaskCreate) -> Task:
        from core.taskiq.tasks import task_process # :)
        task = await self.repo.create_task(task_in=task_in)
        await task_process.kiq(task.id)
        log.info("Task sent to RabbitMQ")
        return task
