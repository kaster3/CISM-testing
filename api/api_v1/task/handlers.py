from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from core.database.models import Task
from core.database.models.enums.task_status_enum import TaskStatusEnum

from .dependencies import get_task_by_id, get_task_service
from .schemas import TaskCreate, TaskSchema
from .service import TaskService

router = APIRouter(
    prefix="/tasks",
    tags=["Task"],
)


@router.get(
    "/{task_id}",
    response_model=TaskSchema,
    status_code=status.HTTP_200_OK,
)
async def get_task_by_id(
    task: Annotated[Task, Depends(get_task_by_id)],
):
    return task


@router.get(
    "/",
    response_model=list[TaskSchema],
    status_code=status.HTTP_200_OK,
)
async def get_tasks(
    service: Annotated[TaskService, Depends(get_task_service)],
    task_status: TaskStatusEnum = Query(None),
):
    tasks = await service.get_tasks(task_status=task_status)
    return tasks


@router.post(
    "/",
    response_model=TaskCreate,
    status_code=status.HTTP_201_CREATED,
)
async def create_task(
    task_in: TaskCreate,
    service: Annotated[TaskService, Depends(get_task_service)],
):
    task = await service.create_task(task_in=task_in)
    return task
