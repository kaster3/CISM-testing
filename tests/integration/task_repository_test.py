import pytest
from syrupy import SnapshotAssertion, snapshot  # noqa: F401

from api.api_v1.task.repository import TaskRepository
from api.api_v1.task.schemas import TaskCreate, TaskUpdatePartial
from core.database.models import Task
from core.database.models.enums.task_status_enum import TaskStatusEnum


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "task_id, expected",
    [
        (1, True),
        (2, True),
        (3, True),
        (4, False),
    ],
)
async def test_get_task_by_id(
    init_models: None,
    create_tasks: None,
    task_id: int,
    expected: bool,
    task_repository: TaskRepository,
    snapshot: SnapshotAssertion,  # noqa: 811
):
    task = await task_repository.get_task(task_id=task_id)

    if expected:
        assert task is not None
        snapshot.assert_match(task)
    else:
        assert task is None


@pytest.mark.asyncio
async def test_get_all_tasks(
    init_models: None,
    create_tasks: None,
    task_repository: TaskRepository,
    snapshot: SnapshotAssertion,  # noqa: 811
):
    tasks = await task_repository.get_all_tasks()
    snapshot.assert_match(tasks)


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "status, quantity",
    [
        (TaskStatusEnum.NEW, 2),
        (TaskStatusEnum.IN_PROGRESS, 1),
        (TaskStatusEnum.COMPLETED, 0),
        (TaskStatusEnum.ERROR, 0),
    ],
)
async def test_get_tasks_by_status(
    init_models: None,
    create_tasks: None,
    status: TaskStatusEnum,
    quantity: int,
    task_repository: TaskRepository,
    snapshot: SnapshotAssertion,  # noqa: 811
):
    tasks = await task_repository.get_tasks_by_status(task_status=status)
    assert len(tasks) == quantity
    snapshot.assert_match(tasks)


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "task_id, task_data",
    [
        (1, TaskCreate(name="Test 1")),
        (2, TaskCreate(name="Test 2")),
    ],
)
async def test_create_task(
    init_models: None,
    task_id: int,
    task_data: TaskCreate,
    task_repository: TaskRepository,
    snapshot: SnapshotAssertion,  # noqa: 811
):

    await task_repository.create_task(task_data)
    task = await task_repository.get_task(task_id=task_id)
    snapshot.assert_match(task)


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "task_data",
    [
        TaskUpdatePartial(name="Test 1", status=TaskStatusEnum.IN_PROGRESS),
        TaskUpdatePartial(name="Test 2", status=TaskStatusEnum.ERROR),
    ],
)
async def test_update_task(
    init_models: None,
    get_task_by_id: Task,
    task_data: TaskUpdatePartial,
    task_repository: TaskRepository,
    snapshot: SnapshotAssertion,  # noqa: 811
):

    updated_task = await task_repository.update_task(
        task=get_task_by_id,
        task_in=task_data,
    )

    task = await task_repository.get_task(task_id=updated_task.id)
    snapshot.assert_match(task)
