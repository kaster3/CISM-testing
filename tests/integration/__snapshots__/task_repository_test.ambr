# serializer version: 1
# name: test_create_task[1-task_data0]
  Task(id=1, name='Test 1', status=TaskStatusEnum.NEW)
# ---
# name: test_create_task[2-task_data1]
  None
# ---
# name: test_get_all_tasks
  list([
    Task(id=1, name='task1', status=TaskStatusEnum.NEW),
    Task(id=2, name='task2', status=TaskStatusEnum.NEW),
    Task(id=3, name='task3', status=TaskStatusEnum.IN_PROGRESS),
  ])
# ---
# name: test_get_task_by_id[1-True]
  Task(id=1, name='task1', status=TaskStatusEnum.NEW)
# ---
# name: test_get_task_by_id[2-True]
  Task(id=2, name='task2', status=TaskStatusEnum.NEW)
# ---
# name: test_get_task_by_id[3-True]
  Task(id=3, name='task3', status=TaskStatusEnum.IN_PROGRESS)
# ---
# name: test_get_tasks_by_status[TaskStatusEnum.COMPLETED-0]
  list([
  ])
# ---
# name: test_get_tasks_by_status[TaskStatusEnum.ERROR-0]
  list([
  ])
# ---
# name: test_get_tasks_by_status[TaskStatusEnum.IN_PROGRESS-1]
  list([
    Task(id=3, name='task3', status=TaskStatusEnum.IN_PROGRESS),
  ])
# ---
# name: test_get_tasks_by_status[TaskStatusEnum.NEW-2]
  list([
    Task(id=1, name='task1', status=TaskStatusEnum.NEW),
    Task(id=2, name='task2', status=TaskStatusEnum.NEW),
  ])
# ---
# name: test_update_task[1-task_data0]
  Task(id=1, name='Test 1', status=TaskStatusEnum.IN_PROGRESS)
# ---
# name: test_update_task[1-task_data1]
  Task(id=1, name='Test 2', status=TaskStatusEnum.ERROR)
# ---
# name: test_update_task[2-task_data0]
  Task(id=2, name='Test 1', status=TaskStatusEnum.IN_PROGRESS)
# ---
# name: test_update_task[2-task_data1]
  Task(id=2, name='Test 2', status=TaskStatusEnum.ERROR)
# ---
